dpctl/memory/_sycl_usm_array_interface_utils.pxi

```diff
@@ -0,0 +1,187 @@
+
+
+cdef bint _valid_usm_ptr_and_context(DPCTLSyclUSMRef ptr, SyclContext ctx):
+    usm_type = _Memory.get_pointer_type(ptr, ctx)
+    return usm_type in (b'shared', b'device', b'host')
+
+
+cdef DPCTLSyclQueueRef _queue_ref_copy_from_SyclQueue(
+    DPCTLSyclUSMRef ptr, SyclQueue q):
+    """ Check that USM ptr is consistent with SYCL context in the queue,
+    and return a copy of QueueRef if so, or NULL otherwise.
+    """
+    cdef SyclContext ctx = q.get_sycl_context()
+    if (_valid_usm_ptr_and_context(ptr, ctx)):
+        return DPCTLQueue_Copy(q.get_queue_ref())
+    else:
+        return NULL
+
+
+cdef DPCTLSyclQueueRef _queue_ref_copy_from_USMRef_and_SyclContext(
+    DPCTLSyclUSMRef ptr, SyclContext ctx):
+    """ Obtain device from pointer and sycl context, use
+        context and device to create a queue from which this memory
+        can be accessible.
+    """
+    cdef SyclDevice dev = _Memory.get_pointer_device(ptr, ctx)
+    cdef DPCTLSyclContextRef CRef = ctx.get_context_ref()
+    cdef DPCTLSyclDeviceRef DRef = dev.get_device_ref()
+    return DPCTLQueue_Create(CRef, DRef, NULL, 0)
```

Comment:
The `CRef` and `DRef` attributes to `DPCTLQueue_Create` have `__dpctl_keep` ownership attribute and should be freed in this function.

Reply:
But `ctx` and `dev` are python objects, and they will make sure `sycl` objects referenced by `DRef` and `CRef` are disposed of

Reply:
Right. `get_context_ref` and `get_device_ref` will give the raw pointer with lifetime managed by the Python objects.